modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/Cycles.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.network.serialization.marshal;
+
+import org.apache.ignite.internal.network.serialization.ClassDescriptor;
+
+/**
+ * Logic related to cycles in object graphs.
+ */
+class Cycles {
+    private final BuiltInNonContainerMarshallers builtInNonContainerMarshallers;
+
+    Cycles(BuiltInNonContainerMarshallers builtInNonContainerMarshallers) {
+        this.builtInNonContainerMarshallers = builtInNonContainerMarshallers;
+    }
+
+    /**
+     * Returns {@code true} if an instance of the type represented by the descriptor may actively form a cycle.
+     *
+     * @param descriptor    descriptor to check
+     * @return {@code true} if an instance of the type represented by the descriptor may actively form a cycle
+     */
+    boolean canParticipateInCycles(ClassDescriptor descriptor) {
+        return !builtInNonContainerMarshallers.supports(descriptor.clazz());
```

Comment:
So any non-builtin type is a type that can form a cycle, right?

Reply:
Currently, it's any type that is not a built-in non-container type. Later, as an optimization, we could widen the notion of non-cycleable types (for example, if a class only contains primitives, it cannot participate in a cycle).